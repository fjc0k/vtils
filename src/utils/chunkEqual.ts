import { chunk } from 'lodash-uni'

/**
 * 类似 `chunk`，但当每组条目数量不一致时会在最后一组添加填充值以达到每组条目数量一致。
 *
 * @param array 数组
 * @param size 分组大小
 * @param filler 填充值
 */
export function chunkEqual<T>(
  array: T[],
  size: number,
  filler: (index: number) => T,
): T[][] {
  const len = array.length
  const remain = len % size
  if (remain !== 0 && size < array.length) {
    array = array.slice()
    for (let i = 0, n = size - remain; i < n; i++) {
      array.push(filler(len + i))
    }
  }
  return chunk(array, size)
}
