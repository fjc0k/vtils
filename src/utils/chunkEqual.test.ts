import { chunkEqual } from './chunkEqual'

describe('chunkEqual', () => {
  test('表现正常', () => {
    expect(chunkEqual([], 3, () => 100)).toMatchSnapshot()
    expect(chunkEqual([1, 2], 3, () => 100)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3], 3, () => 100)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3, 4], 3, () => 100)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3, 4, 5], 3, i => i)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3, 4, 5], 1, i => i)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3, 4, 5], 2, i => i)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3, 4, 5], 8, i => i)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3, 4, 5], 4, i => i)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3, 4], 2, () => 100)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3, 4, 5, 6], 3, () => 100)).toMatchSnapshot()
    expect(chunkEqual([1, 2, 3, 4, 5, 6], 2, () => 100)).toMatchSnapshot()
  })

  test('consistent', () => {
    expect(chunkEqual([1, 2], 3, () => 100, true)).toMatchSnapshot()
  })
})
