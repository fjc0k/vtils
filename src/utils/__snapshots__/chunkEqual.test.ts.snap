// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chunkEqual 表现正常 1`] = `Array []`;

exports[`chunkEqual 表现正常 2`] = `
Array [
  Array [
    1,
    2,
  ],
]
`;

exports[`chunkEqual 表现正常 3`] = `
Array [
  Array [
    1,
    2,
    3,
  ],
]
`;

exports[`chunkEqual 表现正常 4`] = `
Array [
  Array [
    1,
    2,
    3,
  ],
  Array [
    4,
    100,
    100,
  ],
]
`;

exports[`chunkEqual 表现正常 5`] = `
Array [
  Array [
    1,
    2,
    3,
  ],
  Array [
    4,
    5,
    5,
  ],
]
`;

exports[`chunkEqual 表现正常 6`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
  Array [
    4,
  ],
  Array [
    5,
  ],
]
`;

exports[`chunkEqual 表现正常 7`] = `
Array [
  Array [
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
    5,
  ],
]
`;

exports[`chunkEqual 表现正常 8`] = `
Array [
  Array [
    1,
    2,
    3,
    4,
    5,
  ],
]
`;

exports[`chunkEqual 表现正常 9`] = `
Array [
  Array [
    1,
    2,
    3,
    4,
  ],
  Array [
    5,
    5,
    6,
    7,
  ],
]
`;

exports[`chunkEqual 表现正常 10`] = `
Array [
  Array [
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
]
`;

exports[`chunkEqual 表现正常 11`] = `
Array [
  Array [
    1,
    2,
    3,
  ],
  Array [
    4,
    5,
    6,
  ],
]
`;

exports[`chunkEqual 表现正常 12`] = `
Array [
  Array [
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
    6,
  ],
]
`;
